Add maxsat blog and folder scaffolding (#15)
rancher-sandbox/hypper.io: content/blog/2021/08/using-maxsat-for-resolving-deps.md

> # Using a MaxSAT for resolving Hypper chart dependencies on Kubernetes

Reviewer: this sounds like as if "Hypper chart" is something different from "Helm chart".

How about
- Using a MaxSAT for resolving Helm chart dependencies on Kubernetes
or
- Hypper: Using a MaxSAT for resolving chart dependencies on Kubernetes

Author: I like the second!

Reviewer 2: The second sounds good, first one gives the impression that hypper uses helm charts....

Author: Changed it to the second in https://github.com/rancher-sandbox/hypper.io/pull/15/commits/45469e344ceed720af1e49b60c10097fb65bf848.

[Hypper](http://hypper.io) is a Kubernetes package manager based on Helm. It
handles charts in a similar way as Helm does, with a twist: it tries to resolve
dependencies between charts to reuse existing resources and deployments.
For more info, see the two articles on this blog by @mattfarina: [Introducing
Hypper](https://community.suse.com/posts/15893197) and
[Hypper v3 - Now with a SAT solver](https://community.suse.com/posts/14072167).

Hypper is designed to work with shared dependencies: dependencies installed once
on a cluster that several charts depend on. Think of Hypper as bringing the
usual Linux package management to Kubernetes Helm charts.

Checking all the charts so see if their dependencies are correctly installed or

> there's some missing is not an easy task. This is called a "Satisfiability

Reviewer: there are (it's multiple dependencies)

Reviewer 2: Also if we publish on hypper.io first we need to update the references to the posts from @mattfarina to the suse community blog.

Author: Couldn't we keep the existing references to the existing posts from Matt as they keep being there? I would not like to migrate all past posts to a hypper.io/blog.

Reviewer 2: That was not my point, I just meant we should say explicitly in the references to the previous post, that those are hosted on the community blog (to avoid confusion for the reader).

> problem", or SAT, for friends. In fact, solving a SAT problem it's an [NP-Complete

Reviewer: ```suggestion
problem", or SAT, for friends. In fact, solving a SAT problem is a [NP-Complete
```

Author: ahá! :) In theory it should be "a NP problem", because it's a consonant instead of a vowel, but seems that consonants with vowel sounds include f, h, l, m, n, r, s, and x, particularly in acronyms and so. Examples:
- An NBC television show.
- An HTML document.
- A CBS news outlet.
- A Microsoft program, but An MS-DOS program.

Will leave it as it is :).

Reviewer 3: This did not read smoothly to me. It would read more smoothly as "solving a SAT problem is an NP-Complete problem"

The [language to use here is not straight forward](https://english.stackexchange.com/questions/1016/do-you-use-a-or-an-before-acronyms-initialisms). "is an" reads more clearly.

problem](https://en.wikipedia.org/wiki/Boolean_satisfiability_problem#Unrestricted_satisfiability_(SAT))!.

## How to solve a SAT problem?

Finding a solution to that problem may be somewhat easy by reducing the problem
space (like Go modules), or using custom heuristics to be fast but without an
assurance of finding an optimal solution (like Apt).

To find an optimal solution to a SAT problem fast enough so it's useful to use,
we need a focused approach. That comes from a SAT solver algorithm: it gets fed
with facts of a problem, and it finds the optimal combination of facts that can
satisfy the problem or a demonstration that such combination doesn't exist.

In Hypper's case, we also want the possibility to depend on Helm charts outside
of our curated repositories. Therefore, we would benefit from knowing if a
system can be upgraded or not, regardless of our ability on curating the
repositories. Hence, we use a SAT solver directly when performing those
upgrades.

> ## What kind of SAT solvers exist?

Reviewer: ```suggestion
## What kinds of SAT solvers exist?
```

🤔

Author: Fixed in https://github.com/rancher-sandbox/hypper.io/pull/15/commits/45469e344ceed720af1e49b60c10097fb65bf848!

# Hypper: using a MaxSAT for resolving chart dependencies on Kubernetes

[Hypper](http://hypper.io) is a Kubernetes package manager based on Helm. It
handles charts in a similar way as Helm does, with a twist: it tries to resolve
dependencies between charts to reuse existing resources and deployments.
For more info, see the two articles on this blog by @mattfarina: [Introducing
Hypper](https://community.suse.com/posts/15893197) and
[Hypper v3 - Now with a SAT solver](https://community.suse.com/posts/14072167).

Hypper is designed to work with shared dependencies: dependencies installed once
on a cluster that several charts depend on. Think of Hypper as bringing the
usual Linux package management to Kubernetes Helm charts.

Checking all the charts to see if their dependencies are correctly installed or
there are some missing is not an easy task. This is called a "Satisfiability
problem", or SAT, for friends. In fact, solving a SAT problem it's an [NP-Complete
problem](https://en.wikipedia.org/wiki/Boolean_satisfiability_problem#Unrestricted_satisfiability_(SAT))!.

## How to solve a SAT problem?

Finding a solution to that problem may be somewhat easy by reducing the problem

> space (like Go modules), or using custom heuristics to be fast but without an

Reviewer 3: You might want to link to "Go modules". Many people who read this may not know Go or even what Go modules are. I would use the URL https://golang.org/ref/mod.

Reviewer 2: Resolved with 8ceaae9
